core/src/qos_service.rs

```diff
@@ -0,0 +1,242 @@
+//! Quality of service for block producer.
+//! Provides logic and functions to allow a Leader to prioritize
+//! how transactions are included in blocks, and optimize those blocks.
+//!
+use {
+    solana_measure::measure::Measure,
+    solana_runtime::{
+        bank::Bank,
+        cost_model::{CostModel, TransactionCost},
+        cost_tracker::CostTrackerError,
+    },
+    solana_sdk::transaction::{self, SanitizedTransaction, TransactionError},
+    std::sync::{Arc, RwLock},
+};
+
+#[derive(Default)]
+pub struct QosServiceStats {
+    compute_cost_time: u64,
+    cost_tracking_time: u64,
+    selected_txs_count: u64,
+    retried_txs_per_block_limit_count: u64,
+    retried_txs_per_account_limit_count: u64,
+}
+
+impl QosServiceStats {
+    pub fn report(&mut self) {
+        datapoint_info!(
+            "qos-service-stats",
+            ("compute_cost_time", self.compute_cost_time, i64),
+            ("cost_tracking_time", self.cost_tracking_time, i64),
+            ("selected_txs_count", self.selected_txs_count, i64),
+            (
+                "retried_txs_per_block_limit_count",
+                self.retried_txs_per_block_limit_count,
+                i64
+            ),
+            (
+                "retried_txs_per_account_limit_count",
+                self.retried_txs_per_account_limit_count,
+                i64
+            ),
+        );
+    }
+}
+
+pub struct QosService {
+    cost_model: Arc<RwLock<CostModel>>,
```

Comment:
So is this `RwLock` wrapped so that the banking stage's qos service always has the latest on-chain program timings from replay stage? Do you mind adding a comment for that?

Reply:
yep, comments added in #21406